File: sort/radix_sort.cpp
// +build ignore
// from golang

#include "radix_sort.h"
#include <iostream>

unsigned int getDigit(unsigned int value, int bit) {
    unsigned  int result;
    while(bit--) {
        result = value % 10;
        value /= 10;
    }
    return result;
} 

// radix sort - asc
std::vector<unsigned int> radixSort(std::vector<unsigned int> inArr) {
    int size = inArr.size();
    if (size <= 1) return inArr;
    // find max digit
    unsigned int maxDigit = 0;
    for(int i=0; i<size; ++i) 
        if (maxDigit < inArr[i]) maxDigit = inArr[i];
    int flag = 0;
    while(maxDigit > 0) {
        maxDigit /= 10;
        ++flag;
    }
    maxDigit = flag;
    // sort
    unsigned int d = 1;
    while(d <= maxDigit) {
        std::vector<std::vector<unsigned int>> tmpArr(10);
        for(int i=0; i<int(inArr.size()); ++i) tmpArr[getDigit(inArr[i], d)].push_back(inArr[i]);
        inArr.clear();
        for(int i=0; i<10; ++i) {
            for(int z=0; z<int(tmpArr[i].size()); ++z) inArr.push_back(tmpArr[i][z]);
        }
        ++d;
    }
    return inArr;
}

File: sort/radix_sort.go
package sort

func getDigit(value, bit uint) (result uint) {
	for ; bit > 0; bit-- {
		result = value % 10
		value /= 10
	}
	return
}

// radix sort - asc
func RadixSort(inArr []uint) []uint {
	l := len(inArr)
	if l <= 1 {
		return inArr
	}
	// find max digit
	var maxDigit uint
	for i := 0; i < l; i++ {
		if maxDigit < inArr[i] {
			maxDigit = inArr[i]
		}
	}
	var flag uint
	for maxDigit > 0 {
		maxDigit /= 10
		flag++
	}
	maxDigit = flag
	// sort
	var d uint = 1
	for d <= maxDigit {
		var tmpArr [10][]uint
		for i := 0; i < l; i++ {
			tmpArr[getDigit(inArr[i], d)] = append(tmpArr[getDigit(inArr[i], d)], inArr[i])
		}
		inArr = make([]uint, 0)
		for i := 0; i < 10; i++ {
			for z := 0; z < len(tmpArr[i]); z++ {
				inArr = append(inArr, tmpArr[i][z])
			}
		}
		d++
	}
	return inArr
}
